src/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/InteropHelpers.cs

```diff
@@ -71,45 +72,84 @@ internal static unsafe IntPtr ResolvePInvokeSlow(MethodFixupCell* pCell)
             return pCell->Target;
         }
 
-        internal static unsafe void FixupModuleCell(ModuleFixupCell* pCell)
+        #if PLATFORM_UNIX
+            const string PAL_SHLIB_PREFIX = "lib";
+            #if PLATFORM_OSX
+                const string PAL_SHLIB_SUFFIX = ".dylib";
+            #else
+                const string PAL_SHLIB_SUFFIX = ".so";
+            #endif
+        #endif
+        
+        internal static unsafe IntPtr TryResolveModule(string moduleName)
         {
-#if !PLATFORM_UNIX
-            char* moduleName = (char*)pCell->ModuleName;
+            IntPtr hModule = IntPtr.Zero;
 
-            IntPtr hModule = Interop.mincore.LoadLibraryEx(moduleName, IntPtr.Zero, 0);
-            if (hModule != IntPtr.Zero)
-            {
-                var oldValue = Interlocked.CompareExchange(ref pCell->Handle, hModule, IntPtr.Zero);
-                if (oldValue != IntPtr.Zero)
-                {
-                    // Some other thread won the race to fix it up.
-                    Interop.mincore.FreeLibrary(hModule);
-                }
-            }
-            else
-            {
-                // TODO: should be DllNotFoundException, but layering...
-                throw new TypeLoadException(new string(moduleName));
-            }
+            // Try original name first
+            hModule = LoadLibrary(moduleName);
+            if (hModule != IntPtr.Zero) return hModule;
+
+#if PLATFORM_UNIX
+            // Try prefix+name+suffix
+            hModule = LoadLibrary(PAL_SHLIB_PREFIX + moduleName + PAL_SHLIB_SUFFIX);
+            if (hModule != IntPtr.Zero) return hModule;
+
+            // Try name+suffix
+            hModule = LoadLibrary(moduleName + PAL_SHLIB_SUFFIX);
+            if (hModule != IntPtr.Zero) return hModule;
+
+            // Try prefix+name
+            hModule = LoadLibrary(PAL_SHLIB_PREFIX + moduleName);
+            if (hModule != IntPtr.Zero) return hModule;
+#endif
+            return IntPtr.Zero;
+        }
+
+        internal static unsafe IntPtr LoadLibrary(string moduleName)
+        {
+            IntPtr hModule;
+
+#if !PLATFORM_UNIX
+            fixed (char *name = moduleName)
+                hModule = Interop.mincore.LoadLibraryEx(name, IntPtr.Zero, 0);
 #else
-            byte* moduleName = (byte*)pCell->ModuleName;
+            Utf8String utf8ModuleName = new Utf8String(moduleName);
+            fixed (byte *name = utf8ModuleName.UnderlyingArray)
+                hModule = Interop.Sys.LoadLibrary(name);
```

Comment:
We should rename the Unix entrypoint to dlopen to match the interop guidelines: the Unix shims should have the same names as the underlying Unix APIs.